python/cudf/bindings/groupby/groupby.pyx

```diff
@@ -0,0 +1,355 @@
+# Copyright (c) 2018, NVIDIA CORPORATION.
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+# Copyright (c) 2018, NVIDIA CORPORATION.
+
+
+import collections
+import numpy as np
+from numbers import Number
+
+from cython.operator cimport dereference as deref
+from libcpp.utility cimport pair
+from libcpp.vector cimport vector
+from libc.stdlib cimport malloc, free
+
+import cudf
+import cudf.dataframe.index as index
+from cudf.bindings.cudf_cpp cimport *
+from cudf.bindings.cudf_cpp import *
+from cudf.dataframe.dataframe import DataFrame
+from cudf.dataframe.series import Series
+from cudf.dataframe.buffer import Buffer
+from cudf.dataframe.categorical import CategoricalColumn
+from cudf.utils.cudautils import zeros
+from cudf.bindings.nvtx import nvtx_range_pop
+from cudf.bindings.utils cimport *
+cimport cudf.bindings.groupby.hash as hash_groupby
+
+
+agg_names = {
+    'sum': hash_groupby.SUM,
+    'min': hash_groupby.MIN,
+    'max': hash_groupby.MAX,
+    'count': hash_groupby.COUNT,
+    'mean': hash_groupby.MEAN
+}
+
+
+def columns_from_dataframe(df):
+    return [sr._column for sr in df._cols.values()]
+
+
+def dataframe_from_columns(cols, index=None, columns=None):
+    df = cudf.DataFrame(dict(zip(range(len(cols)), cols)))
+    if columns is not None:
+        df.columns = columns
+    return df
+
+
+def apply_groupby(keys, values, ops, method='hash'):
+    """
+    Apply aggregations *ops* on *values*, grouping by *keys*.
+
+    Parameters
+    ----------
+    keys : list of Columns
+    values : list of Columns
+    ops : str or list of str
+        Aggregation to be performed for each column in *values*
+
+    Returns
+    -------
+    result : tuple of list of Columns
+        keys and values of the result
+    """
+    if len(values) == 0:
+        return (keys, [])
+
+    cdef pair[cudf_table, cudf_table] result
+    cdef cudf_table *c_keys_table = table_from_columns(keys)
+    cdef cudf_table *c_values_table = table_from_columns(values)
```

Comment:
Leaking these tables as they're being allocated in their util functions with no corresponding del?